### PR TITLE
Create app from template

### PR DESCRIPTION
**Pull request by tgorgol**
**Description**

Changes proposed in this pull request:

- Add mutation to create application from application template
- Add validation to application template domain

**Related issue(s)**

#441

### CODE REVIEW

File: components/director/pkg/graphql/app_template_validation.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	"strings"
+
+	validation "github.com/go-ozzo/ozzo-validation"
+	"github.com/kyma-incubator/compass/components/director/pkg/inputvalidation"
+	"github.com/pkg/errors"
+)
+
+func (i ApplicationTemplateInput) Validate() error {
+	return validation.Errors{
+		"rule.ValidPlaceholders": i.validPlaceholders(),
```
Comment (crabtree) [inline components/director/pkg/graphql/app_template_validation.go]: Should be `rule.ValidPlaceholders` or `Rule.ValidPlaceholders`? In different files, I can see different approaches. (See PR #516)

File: components/director/pkg/graphql/app_template_validation.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+func (i PlaceholderDefinitionInput) Validate() error {
+	return validation.ValidateStruct(&i,
+		validation.Field(&i.Name, validation.Required, inputvalidation.Name),
+		validation.Field(&i.Description, validation.RuneLength(0, shortStringLengthLimit)),
+	)
+}
+
+func (i ApplicationFromTemplateInput) Validate() error {
+	return validation.Errors{
+		"rule.UniquePlaceholders": i.ensureUniquePlaceholders(),
```
Comment (crabtree) [inline components/director/pkg/graphql/app_template_validation.go]: Should be `rule.UniquePlaceholders` or `Rule.UniquePlaceholders`? In different files, I can see different approaches. (See PR #516)

File: components/director/pkg/graphql/app_template_validation.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+		"TemplateName":            validation.Validate(i.TemplateName, validation.Required, inputvalidation.Name),
+		"Values":                  validation.Validate(i.Values, validation.Each(validation.Required)),
+	}.Filter()
+}
+
+func (i ApplicationFromTemplateInput) ensureUniquePlaceholders() error {
```
Comment (crabtree) [inline components/director/pkg/graphql/app_template_validation.go]: In this file, there are two functions that do almost the same but on different types and their names are different. Please make it more consistent.

File: components/director/pkg/graphql/app_template_validation.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+	keys := make(map[string]interface{})
+	for _, item := range i.Values {
+		if item == nil {
+			continue
+		}
+		_, exist := keys[item.Placeholder]
```
Comment (crabtree) [inline components/director/pkg/graphql/app_template_validation.go]:
```go
if _, exist := keys[item.Placeholder]; exist {
	return errors.Errorf("placeholder [name=%s] not unique", item.Placeholder)
}

keys[item.Placeholder] = struct{}{}
```

File: components/director/pkg/graphql/app_template_validation.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+	return nil
+}
+
+func (i ApplicationTemplateInput) ensurePlaceholdersUnique() error {
+	keys := make(map[string]interface{})
+	for _, item := range i.Placeholders {
+		if item == nil {
+			continue
+		}
+		_, exist := keys[item.Name]
```
Comment (crabtree) [inline components/director/pkg/graphql/app_template_validation.go]:
```go
if _, exist := keys[item.Name]; exist {
	return errors.Errorf("placeholder [name=%s] not unique", item.Name)
}

keys[item.Name] = struct{}{}
```

File: components/director/internal/domain/apptemplate/service.go [hunk 6]
```diff
@@ -105,15 +106,23 @@ func (s *service) Delete(ctx context.Context, id string) error {
...
+		newValue, err := s.lookForPlaceholderValue(templatePlaceholderValues, placeholder.Name)
+		if err != nil {
+			return "", err
+		}
+		appCreateInputJSON = strings.ReplaceAll(appCreateInputJSON, fmt.Sprintf("{{%s}}", placeholder.Name), newValue)
+	}
+	return appCreateInputJSON, nil
+}
+
+func (s *service) lookForPlaceholderValue(values []*model.ApplicationTemplateValueInput, placeholderName string) (string, error) {
```
[the remark this answers is not in the document]
Comment (pkosiec) [reply components/director/internal/domain/apptemplate/service.go]: Good idea, but I would need to go through all layers to introduce the new `Values` type, check for conversion errors, etc. I think this additional type doesn't bring that much value, to make this effort.

Instead, I've added a method to `ApplicationFromTemplateInput` itself.

[the remark this answers is not in the document]
Comment (pkosiec) [reply components/director/internal/domain/apptemplate/service.go]: good idea - done

**Pull request merged**